ddtrace/contrib/wsgi/wsgi.py

```diff
@@ -75,36 +113,38 @@ def _response_span_name(self):
         raise NotImplementedError
 
     def __call__(self, environ, start_response):
-        # type: (Iterable, Callable) -> None
+        # type: (Iterable, Callable) -> _TracedIterable
         trace_utils.activate_distributed_headers(self.tracer, int_config=self._config, request_headers=environ)
 
-        with self.tracer.trace(
+        req_span = self.tracer.trace(
             self._request_span_name,
             service=trace_utils.int_service(self._pin, self._config),
             span_type=SpanTypes.WEB,
-        ) as req_span:
-            # This prevents GeneratorExit exceptions from being set on the request span.
-            req_span._ignore_exception(GeneratorExit)
-            self._request_span_modifier(req_span, environ)
-
-            with self.tracer.trace(self._application_span_name) as app_span:
-                intercept_start_response = functools.partial(self._traced_start_response, start_response, req_span)
-                result = self.app(environ, intercept_start_response)
-                self._application_span_modifier(app_span, environ, result)
-
-            with self.tracer.trace(self._response_span_name) as resp_span:
-                # This prevents GeneratorExit exceptions from being set on the response span.
-                # This can occur if a streaming response exits abruptly leading to a broken pipe.
-                resp_span._ignore_exception(GeneratorExit)
-                self._response_span_modifier(resp_span, result)
-                for chunk in result:
-                    yield chunk
-
-            if hasattr(result, "close"):
-                result.close()
-
-    def _traced_start_response(self, start_response, request_span, status, environ, exc_info=None):
-        # type: (Callable, Span, str, Dict, Any) -> None
+        )
+        self._request_span_modifier(req_span, environ)
+
+        try:
+            app_span = self.tracer.start_span(self._application_span_name, child_of=req_span, activate=True)
+            intercept_start_response = functools.partial(
+                self._traced_start_response, start_response, req_span, app_span
+            )
+            result = self.app(environ, intercept_start_response)
+            self._application_span_modifier(app_span, environ, result)
+            app_span.finish()
+        except Exception:
+            req_span.set_exc_info(*sys.exc_info())
+            app_span.set_exc_info(*sys.exc_info())
+            app_span.finish()
+            req_span.finish()
+            raise
+
```

Comment:
worth having a comment here to explain why we manually do the parenting and then activate=True